Add working schedule resolvers

zacharyzollman opened the pull request:
### Functionality
This adds the capability to create, edit, delete, and retrieve schedules. In addition to retrieving by ObjectID, there's also an option to retrieve all schedules created by a particular user.

### Future changes
Some aspects of this are likely to change. In particular, I plan to add an argument for custom events to the resolvers for creating and editing schedules. I also plan to integrate this with our user authentication system, once that is established. The resolvers for updating specific elements of a schedule may not be helpful for integration with frontend, so I might remove them. I might also tweak the function names.

### Note on local testing
In order for this to work locally, I think you need to add a Collection named "schedule" to the local MongoDB.

maxmwang commented on src/modules/schedule/controller.ts:
```diff
@@ -0,0 +1,115 @@
+// create a new schedule
+export async function createSchedule(created_by: string, term: string, schedule_name: string | undefined | null, is_public: boolean, class_IDs: string[] | undefined | null, primary_section_IDs: string[], secondary_section_IDs: string[]): Promise<Schedule> {
+  
+  const newID = new ObjectID()
```
Instead of having to do `new ObjectID()` for every new document, should let mongoose handle it by not specifying `_id` in `model.ts`. For types, extend the model type with Document.

maxmwang commented on src/modules/schedule/controller.ts:
```diff
@@ -0,0 +1,115 @@
+  
+  const newID = new ObjectID()
+
+  const schedulePartsToCreate: minimumViableSchedule = { _id: newID, created_by: created_by, term: term }
+  if (schedule_name) {
+    schedulePartsToCreate.name = schedule_name
+  }
+
+  if (class_IDs) {
+    schedulePartsToCreate.class_IDs = class_IDs
+  }
+
+  if (primary_section_IDs) {
+    schedulePartsToCreate.primary_section_IDs = primary_section_IDs
+  }
+
+  if (secondary_section_IDs) {
+    schedulePartsToCreate.secondary_section_IDs = secondary_section_IDs
+  }
+
+  schedulePartsToCreate.public = is_public
+
+  const newSchedule = await ScheduleModel.create(schedulePartsToCreate)
+  // const newSchedule = await ScheduleModel.create({_id: newID, name: schedule_name, created_by: created_by, date_created: current_time, last_updated: current_time, term: term, public: is_public, class_IDs: class_IDs, section_IDs: section_IDs})
```
I think this would be much more readable if it had ternary operators instead of if statements. Something like 

```ts
{
  name: schedule_name ? schedule_name : undefined,
  class_IDs: class_IDs ? class_IDs : undefined,
}
```

maxmwang commented on src/modules/schedule/model.ts:
```diff
@@ -0,0 +1,28 @@
+export const CustomEventModel = mongoose.model("customEvent", CustomEventSchema, "customEvent");
+export type CustomEventType = InferSchemaType<typeof CustomEventSchema>;
+export const ScheduleModel = mongoose.model("schedule", ScheduleSchema, "schedule");
+export type ScheduleType = InferSchemaType<typeof ScheduleSchema>;
```
```ts
export type ScheduleType = Document & InferSchemaType<typeof ScheduleSchema>
```
With mongoose@6.8, timestamps are also included in typing.

maxmwang commented on src/modules/schedule/model.ts:
```diff
@@ -0,0 +1,28 @@
+import mongoose, { Schema, InferSchemaType, Types } from "mongoose";
+
+export const CustomEventSchema = new Schema({
+  start_time: { type: String, required: true },
+  end_time: { type: String, required: true },
+  title: { type: String, required: false },
+  location: { type: String, required: false },
+  description: { type: String, required: false },
+  days_of_week: { type: String, required: false }
+});
+
+
+export const ScheduleSchema = new Schema({
+  _id: { type: Types.ObjectId, reqired: true },
+  name: { type: String, required: false },
+  created_by: { type: String, required: true },
+  term: { type: String, required: true },
+  public: { type: Boolean, required: true },
+  class_IDs: { type: [String], required: false },
+  primary_section_IDs: { type: [String], required: false },
+  secondary_section_IDs: { type: [String], required: false },
+  custom_events: { type: [CustomEventSchema], required: false }
```
Remove _id. 

Might be good to consider defaults and other [options](https://mongoosejs.com/docs/schematypes.html#schematype-options).

maxmwang commented on src/modules/schedule/typedefs/schedule.ts:
```diff
@@ -0,0 +1,87 @@
+    """
+    Identifiers (probably cs-course-ids) for the classes the user has added to their schedule.
+    """
+    class_IDs: [String]
```
`[String!]` ensures that all entries of array are non-null strings—should be useful for most `[String]`

maxmwang commented on src/modules/schedule/controller.ts:
```diff
@@ -0,0 +1,115 @@
+// update an existing schedule
+export async function editSchedule(schedule_ID: string, created_by: string | undefined | null, term: string | undefined | null, schedule_name: string | undefined | null, class_IDs: string[], primary_section_IDs: string[], secondary_section_IDs: string[], is_public?: boolean | undefined | null): Promise<Schedule> {
+  
+  const schedulePartsToUpdate: partialSchedule = {}
+
+  if (term) {
+    schedulePartsToUpdate.term = term
+  }
+  if (schedule_name) {
+    schedulePartsToUpdate.name = schedule_name
+  }
+  if (class_IDs) {
+    schedulePartsToUpdate.class_IDs = class_IDs
+  }
+  if (primary_section_IDs) {
+    schedulePartsToUpdate.primary_section_IDs = primary_section_IDs
+  }
+
+  if (secondary_section_IDs) {
+    schedulePartsToUpdate.secondary_section_IDs = secondary_section_IDs
+  }
+
+  if (isBoolean(is_public)) {
+    schedulePartsToUpdate.public = is_public
+  }
```
Same as before.

maxmwang commented on src/modules/schedule/controller.ts:
```diff
@@ -0,0 +1,115 @@
+  if (!scheduleFromID) {
+    throw new Error("No schedules found with this ID")
+  }
+  return formatSchedule(scheduleFromID as any)
```
Wondering if the `as any` is necessary?

zacharyzollman commented:
I think I addressed all of the above points. I also made the term argument an object in order to be more consistent with catalog-related queries. Then I felt it was inconsistent to not return a comparable object, so I updated it so that term is now inputted, stored, and returned as an object.

maxmwang commented on src/modules/schedule/typedefs/schedule.ts:
```diff
@@ -1,6 +1,13 @@
 
 const typedef = gql`
+
+  type TermOutput {
+    year: Int!
+    semester: String!
+  }
```
Is it possible to just use `TermInput`? I know that `TermInput` uses an enum but shouldn't the values of the enum be strings already?

zacharyzollman replied:
I initially tried using the term `input` as both input and output, but I was running into issues and eventually concluded that trying to use an input type as an output was the cause. [These](https://stackoverflow.com/questions/41515679/can-you-make-a-graphql-type-both-an-input-and-output-type) [two](https://stackoverflow.com/questions/41743253/whats-the-point-of-input-type-in-graphql) stackoverflow posts seem to confirm that.

maxmwang replied:
Ah I see, quite a shame.

maxmwang commented on src/modules/schedule/controller.ts:
```diff
@@ -35,65 +33,40 @@ export async function removeSchedule(scheduleID: string): Promise<string> {
+  schedulePartsToCreate.secondary_section_IDs = secondary_section_IDs ? secondary_section_IDs : undefined
+  schedulePartsToCreate.is_public = is_public ? is_public : undefined
 
   const newSchedule = await ScheduleModel.create(schedulePartsToCreate)
```
(continued from `resolver.ts` comment)

I know I suggested this, but I still think this is quite hard to work with. Consider the case where we add an extra field to the schedule schema: we would have to come here and refactor this code. 

I played around with how to more easily create (this also applies to update/edit), and passing the object directly into `create()` should work much cleaner:

```ts
// the `createSchedule()` function now takes in one argument `args`

const newSchedule = await ScheduleModel.create(args);
```

Frustratingly, if the graphql inputs can be null (no !), then users can explicitely pass in `null`. Mongoose will not use a default value if `null` is passed in, and so now there is a `null` value in a MongoDB document. When queried, mongoose will give a null value in a field declared non-null by graphql. Not too sure what the best fix for this is, but for now I just loop over the `args` argument and `delete` null fields. 

Edit: as per Kev's suggestion, using `omitBy` from lodash will easily remove explicit null values from an object.

zacharyzollman replied:
Regarding removing explicit null values, my input type is a bit complicated. Within it, I have another input type (the term) and an array of input types (the custom events). I was getting type issues with the return value of `omitBy` so I did some web searches and didn't see an easy solution. For now I adapted the code you initially shared on Discord.

maxmwang replied:
Hm yeah, seems like a recursive `omitBy` doesn't exist.

maxmwang commented on src/modules/schedule/typedefs/schedule.ts:
```diff
@@ -68,11 +75,11 @@ const typedef = gql`
     Takes in schedule fields, creates a new schedule record in the database, and returns the schedule.
     """
-    createNewSchedule(created_by: String!, term: String!, schedule_name: String, is_public: Boolean, class_IDs: [String!], primary_section_IDs: [String!], secondary_section_IDs: [String!]): Schedule
+    createNewSchedule(created_by: String!, term: TermInput!, schedule_name: String, is_public: Boolean, class_IDs: [String!], primary_section_IDs: [String!], secondary_section_IDs: [String!]): Schedule
```
As Kev suggested, consider defining an [`input` type](https://graphql.org/graphql-js/mutations-and-input-types/) for arguments for cleanliness.

zacharyzollman commented:
@maxmwang Thanks for the input! Those are good points, I think it is a lot cleaner now.

maxmwang reviewed (approved):
Nice changes, looks good to me!
